### PR TITLE
PYTHON-4786 - Fix UpdateResult.did_upsert TypeError

### PR DESCRIPTION
**Pull request by NoahStapp**

### CODE REVIEW

**Review by blink1073 — DISMISSED** [review]
LGTM!

File: pymongo/results.py [hunk 1]
```diff
@@ -171,9 +171,10 @@ def upserted_id(self) -> Any:
 
     @property
     def did_upsert(self) -> bool:
-        """Whether or not an upsert took place."""
+        """Whether an upsert took place."""
         assert self.__raw_result is not None
-        return len(self.__raw_result.get("upserted", {})) > 0
+        result = self.__raw_result.get("upserted")
+        return result is not None and len(str(result)) > 0
```
Comment (ShaneHarvey) [inline pymongo/results.py:177]: This doesn't handle the `None` case correctly.

Comment (ShaneHarvey) [reply pymongo/results.py:177]: We need a few more tests. For both client.bulk_write and collection.update_one() that tests upserting a document with `"_id": None`.

Comment (NoahStapp) [reply pymongo/results.py:177]: Do we expect something besides `False` if the result is `None`?

Comment (ShaneHarvey) [reply pymongo/results.py:177]: yes None is a valid id

**Comment by ShaneHarvey** [comment]
For example:
```python
>>> res = client.t.t.update_one({"_id": None}, {'$set': {'x': 1}}, upsert=True)
>>> res
UpdateResult({'n': 1, 'upserted': None, 'nModified': 0, 'ok': 1.0, 'updatedExisting': False}, acknowledged=True)
>>> res.upserted_id
>>> res.upserted_id is None
True
```

**Comment by NoahStapp** [comment]
> For example:
> 
> ```python
> >>> res = client.t.t.update_one({"_id": None}, {'$set': {'x': 1}}, upsert=True)
> >>> res
> UpdateResult({'n': 1, 'upserted': None, 'nModified': 0, 'ok': 1.0, 'updatedExisting': False}, acknowledged=True)
> >>> res.upserted_id
> >>> res.upserted_id is None
> True
> >>> res.did_upsert
> Traceback (most recent call last):
>   File "<stdin>", line 1, in <module>
>   File "/Users/shane/git/mongo-python-driver/pymongo/results.py", line 176, in did_upsert
>     return len(self.__raw_result.get("upserted", {})) > 0
>            ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^
> TypeError: object of type 'NoneType' has no len()
> ```

I can't reproduce this locally, I get this result instead:

```python
>>> from pymongo import MongoClient
>>> client = MongoClient()
>>> res = client.t.t.update_one({"_id": None}, {'$set': {'x': 1}}, upsert=True)
>>> res
UpdateResult({'n': 1, 'nModified': 0, 'ok': 1.0, 'updatedExisting': True}, acknowledged=True)
>>> res.upserted_id
>>> res.upserted_id is None
True
>>> res.did_upsert
False
```

**Comment by ShaneHarvey** [comment]
The problem with your repro attempt is that it did not perform an upsert (see `'updatedExisting': True`) which means your collection already had a document with `_id:None`.

**Comment by NoahStapp** [comment]
> The problem with your repro attempt is that it did not perform an upsert (see `'updatedExisting': True`) which means your collection already had a document with `_id:None`.

Good catch--I tried again with a fresh DB and successfully reproduced an incorrect result without an error:

```python
>>> from pymongo import MongoClient
>>> client = MongoClient()
>>> res = client.t.t.update_one({"_id": None}, {'$set': {'x': 1}}, upsert=True)
>>> res
UpdateResult({'n': 1, 'upserted': None, 'nModified': 0, 'ok': 1.0, 'updatedExisting': False}, acknowledged=True)
>>> res.did_upsert
False
```

I'll add a test to catch this case with a fix.

File: test/test_results.py [hunk 2]
```diff
@@ -122,6 +122,28 @@ def test_update_result(self):
...
+        self.assertEqual(result.did_upsert, True)
+
+        raw_result_3 = {
+            "n": 1,
+            "nModified": 1,
+        }
+        self.repr_test(UpdateResult, raw_result_3)
+
+        result = UpdateResult(raw_result_3, True)
+        self.assertEqual(result.did_upsert, False)
```
Comment (ShaneHarvey) [inline test/test_results.py:146]: This is good but we need integration tests for these cases too.

**Review by ShaneHarvey — CHANGES_REQUESTED** [review]
Please add similar tests for client.bulk_write with verbose=True and coll.update_one.

**Comment by blink1073** [comment]
drivers-pr-bot please backport to v4.9

**Review by blink1073 — DISMISSED** [review]
LGTM!

File: test/test_client_bulk_write.py [hunk 3]
```diff
@@ -550,6 +550,47 @@ def test_returns_error_if_auto_encryption_configured(self):
             "bulk_write does not currently support automatic encryption", context.exception._message
         )
 
+    @client_context.require_version_min(8, 0, 0, -24)
+    @client_context.require_no_serverless
+    @unittest.skipUnless(_HAVE_PYMONGOCRYPT, "pymongocrypt is not installed")
```
Comment (ShaneHarvey) [inline test/test_client_bulk_write.py:555]: Why require `_HAVE_PYMONGOCRYPT`?

Comment (NoahStapp) [reply test/test_client_bulk_write.py:555]: Whoops, artifact of copy-and-paste.

File: pymongo/results.py [hunk 4]
```diff
@@ -171,9 +171,9 @@ def upserted_id(self) -> Any:
 
     @property
     def did_upsert(self) -> bool:
-        """Whether or not an upsert took place."""
+        """Whether an upsert took place."""
```
Comment (ShaneHarvey) [inline pymongo/results.py:174]: Please add `versionadded: 4.9`

**Review by ShaneHarvey — APPROVED** [review]
LGTM once the tests pass.

**Comment by blink1073** [comment]
drivers-pr-bot please backport to v4.9

**Pull request merged**